File: Clients/Go/Socket_client/Data/interfaces/EventSlice.go
package interfaces

type EventString struct {
	Name          string                                                              `json:"name"`
	EventProccess func(event Event, conn *Client, messagePid int, destination string) `json:"eventProccess"`
}

type EventsSubscribed struct {
	Events []string `json:"events"`
}

type EventSlice []EventString

var EventSliceInstance *EventsSubscribed = &EventsSubscribed{Events: []string{}}

func (e *EventSlice) AddEvent(event string, handleFunction func(event Event, conn *Client, messagePid int, destination string)) {
	*e = append(*e, EventString{Name: event, EventProccess: handleFunction})
	EventSliceInstance.Events = append(EventSliceInstance.Events, event)
}

func (e *EventSlice) RemoveEvent(event string) {
	for i, v := range *e {
		if v.Name == event {
			*e = append((*e)[:i], (*e)[i+1:]...)
		}
	}
	for i, v := range EventSliceInstance.Events {
		if v == event {
			EventSliceInstance.Events = append(EventSliceInstance.Events[:i], EventSliceInstance.Events[i+1:]...)
		}
	}
}

func (e *EventSlice) Len() int64 {
	return int64(len(*e))
}

File: Clients/Python/Socket_client/Data/interfaces/EventSlice.py
"""Event registry mirroring the Go EventSlice implementation."""

from __future__ import annotations

import threading
from dataclasses import dataclass
from typing import TYPE_CHECKING, Callable, Dict, Iterable, List

try:
    import socket
except ImportError:  # pragma: no cover
    socket = None  # type: ignore

if socket is None:  # pragma: no cover
    raise RuntimeError("socket module is required")

if TYPE_CHECKING:  # pragma: no cover
    from .Event_State import Event
    from Client import Client


EventHandler = Callable[["Event", "Client", int,str], None]

@dataclass
class EventsSubscribed:
    events: list[str]
    def to_dict(self) -> Dict[str, list[str]]:
        return {"events": self.events}

EventSliceInstance = EventsSubscribed(events=[])

@dataclass(frozen=True)
class EventString:
    name: str
    event_process: EventHandler


class EventSlice:
    """Thread-safe container for custom event handlers."""

    def __init__(self) -> None:
        self._events: List[EventString] = []
        self._lock = threading.RLock()

    def add_event(self, name: str, handler: EventHandler) -> None:
        with self._lock:
            self._events.append(EventString(name=name, event_process=handler))
            EventSliceInstance.events.append(name)

    def remove_event(self, name: str) -> None:
        with self._lock:
            self._events = [evt for evt in self._events if evt.name != name]
            EventSliceInstance.events.remove(name)

    def __len__(self) -> int:
        with self._lock:
            return len(self._events)

    def __iter__(self) -> Iterable[EventString]:
        with self._lock:
            snapshot = list(self._events)
        return iter(snapshot)


__all__ = ["EventSlice", "EventString", "EventHandler"]
